### ISSUE TITLE
Should there be a cudatoolkit-static package?

### ISSUE THREAD

**Issue by gmarkall** [conda-forge/cudatoolkit-feedstock#27]
cc @isuruf

Opening this issue as requested in https://github.com/conda-forge/cudatoolkit-feedstock/issues/15#issuecomment-722911854:

>>>Also why are we distributing ./lib/libcudadevrt.a? There's no need to have a static library at runtime. defaults is not distributing this file.

>> This is needed for JIT compilers such as Numba, which compile and link code at runtime - e.g. it needs to be linked in to any kernel using cooperative grids.

> Then they should go in a separate `cudatoolkit-static` package. Can you please open an issue?

However, defaults distributes `libcudadevrt.a` in the cudatoolkit package, and a later comment states:

>> Anyways I wouldn't get too hung up on what defaults packages of cudatoolkit do or don't include certain files as only the latest version got updates.
> That's just wrong. If we don't include a file that defaults does, then we'll be breaking users who rely on it.

So it also seems that the cudatoolkit package must include `libcudadevrt.a`.

**Comment by leofang** [comment]
CuPy also needs this static library to be included to support dynamic parallelism and cooperative group. It's just a maintenance burden to separate them.

**Comment by jaimergp** [comment]
Can `cudatoolkit` add `cudatoolkit-static` as a runtime dependency?

**Comment by isuruf** [comment]
> CuPy also needs this static library to be included to support dynamic parallelism and cooperative group. It's just a maintenance burden to separate them.

How? `cudatoolkit=10.1` doesn't even have this in `defaults`

> Can cudatoolkit add cudatoolkit-static as a runtime dependency?

What would that achieve?

**Comment by jakirkham** [comment]
> > CuPy also needs this static library to be included to support dynamic parallelism and cooperative group. It's just a maintenance burden to separate them.
> 
> How? `cudatoolkit=10.1` doesn't even have this in `defaults`
> 
> > Can cudatoolkit add cudatoolkit-static as a runtime dependency?
> 
> What would that achieve?

This goes back to the same point I was making before in the other thread. `defaults` doesn't have a way of updating the old `cudatoolkit` versions with the changes in new `cudatoolkit` version. We proposed some options in issue ( https://github.com/AnacondaRecipes/cudatoolkit-feedstock/issues/8 ), but nothing came of it.

As a result old packages in `defaults` lack changes made to new packages  in `defaults`.

> > CuPy also needs this static library to be included to support dynamic parallelism and cooperative group. It's just a maintenance burden to separate them.
> 
> How? `cudatoolkit=10.1` doesn't even have this in `defaults`

Presumably making `cudatoolkit-static` a `run` dependency of `cudatoolkit`. To be clear this is just my interpretation of the conversation that's happened here thus far. Others please feel free to correct me 🙂

**Comment by leofang** [comment]
> > CuPy also needs this static library to be included to support dynamic parallelism and cooperative group. It's just a maintenance burden to separate them.
>
> How? `cudatoolkit=10.1` doesn't even have this in `defaults`

Right, so the packages on `default` are actually broken. The full CuPy test suite would not pass with `default`'s `cudatoolkit`. I've been wanting to fix this (https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5#issuecomment-641309451), but no one from Anaconda responded.

**Comment by jakirkham** [comment]
Just to add some context here (I don't think anyone has brought this up yet), conda-forge has a policy of splitting out static libraries in separate packages per [CFEP-18]( https://github.com/conda-forge/cfep/blob/70f80e245a3497be35cbcf03df91d8b53b4e7931/cfep-18.md ). Hopefully that provides more context around the ask here 🙂

Also another piece of context is `defaults` already ships this library in its `cudatoolkit` package. So keeping it here helps us stay aligned with `defaults`.

Additionally I think it helps for us to look at how conda-forge thinks about static libraries and how people use `libcudadevrt.a`.

The way conda-forge thinks of static libraries is something that would be used during a build only. Given the tooling in conda-forge, dynamic libraries only really make sense for package dependencies since we can track where they are used, create migrators to update to new versions where needed, and also support a range of versions in downstream packages without rebuilding where possible. So this both works well with conda-forge tooling and keeps maintenance manageable.

The use cases of `libcudadevrt.a` that I'm aware of (and others please feel free to add here) are CuPy and Numba, which make use of this library at runtime when generating and compiling kernels. As such this really is a runtime dependency that is needed on a user's system for the libraries to work. It's also not something that CuPy and Numba are linking against during their builds (at least as far as I'm aware). So having new versions of this library don't really change those packages making it more like a dynamic library (at least in usage).

Maybe this helps us bridge a bit of a communication gap (at least as I see it)? 😉 Please feel free to add anything else or contradict me if needed 🙂

**Comment by leofang** [comment]
IMHO libcudadevrt.a is really a runtime thing (hence the name "rt" --- IIRC it's full name is CUDA device runtime library) that for some reason is delivered as a static library. I don't think CFEP-18 should be applied here.

**Comment by isuruf** [comment]
> IMHO libcudadevrt.a is really a runtime thing (hence the name "rt" --- IIRC it's full name is CUDA device runtime library) that for some reason is delivered as a static library. I don't think CFEP-18 should be applied here.

It's not special that way. For eg: some packages require blas headers and libraries at runtime. For those, I recommended adding the specific packages as runtime requirements. It's fine for `cupy` and `numba` to depend on this, but they should depend on this explicitly instead of indirectly via `cudatoolkit`.

**Comment by leofang** [comment]
I'd prefer not to do so, as dynamic parallelism, cooperative groups, etc that CuPy and Numba aim to support are part of the CUDA language. We don't invent these. I think splitting cudadevrt from `cudatoolkit` might cause confusion. At least I would expect installing `cudatoolkit` could bring me most of the things that also come with a runfile install of CTK. Just my two cents.

**Comment by leofang** [comment]
Another problem is the increase of maintenance complexity. On this point I strongly prefer a single recipe to handle all these.

**Comment by isuruf** [comment]
> Another problem is the increase of maintenance complexity. On this point I strongly prefer a single recipe to handle all these.

we can use outputs for these

> I think splitting cudadevrt from cudatoolkit might cause confusion. At least I would expect installing cudatoolkit could bring me most of the things that also come with a runfile install of CTK. Just my two cents.

some packages like 10.1 don't even have these, so there's already confusion. We already decided in CFEP-18 that it's good to separate these out and depend on `-static` when needed.

**Comment by leofang** [comment]
> > Another problem is the increase of maintenance complexity. On this point I strongly prefer a single recipe to handle all these.
> 
> we can use outputs for these

This is also what I'd like to see, but the prerequisite is someone actually worked this out first... We should open a separate issue to discuss the merging strategy for all existing branches, though.

> some packages like 10.1 don't even have these, so there's already confusion. 

As I mentioned earlier (https://github.com/conda-forge/cudatoolkit-feedstock/issues/27#issuecomment-723208100), it was a mistake that should have not happened and have been fixed. This does not justify the status quo. 

> We already decided in CFEP-18 that it's good to separate these out and depend on `-static` when needed.

I have also argued that CFEP-18 does not apply here, so I have no more to say regarding this. If anyone feels strongly about it and has time to invest, I will be happy to review. I listed myself as a maintainer here, but unfortunately I was caught up recently and couldn't help...😞

**Comment by isuruf** [comment]
> I have also argued that CFEP-18 does not apply here, so I have no more to say regarding this.

It does apply here. For eg: `python` and `libpython-static` was split off even though some packages rely on `libpython-static`. Those packages should explicitly depend on those static packages.

**Comment by YeungLy** [comment]
Is `cudatoolkit` at conda including CUDA header files? Since I try to compile some custom extensions and failed. It couldnt find `cuda_runtime_api.h`. I checked at conda path, and can only find libcuda*.so at `anaconda3/pkgs/cudatookit/lib` .

**Comment by jaimergp** [comment]
That's usually provider by the drivers (provided by a Docker image on Linux, the Nvidia installer on Windows; check [here](https://conda-forge.org/docs/maintainer/knowledge_base.html#cuda-builds)).

**Comment by leofang** [comment]
@YeungLy The CUDA headers (`cuda.h`, `cuda_runtime.h`, etc) are not included in `cudatoolkit`, which only includes CUDA runtime libraries. CUDA headers are not redistributable according to NVIDIA's EULA, so if you need them you'll need to use either [NVIDIA-provided docker images](https://hub.docker.com/r/nvidia/cuda) (but not the `runtime` version therein, which is similar to `cudatoolkit` here), or install CUDA Toolkit locally, as they are *not* part of the driver. In short, to compile CUDA code locally you'll need more than `cudatoolkit`. Conda-Forge's toolchain does not support this need.

This is the N-th time we encounter this question (I contributed one 😆). I think it's necessary to add docs to Conda-Forge's user (not maintainer) documentation...

**Comment by leofang** [comment]
(And, for further discussion on the header issue, please open a new issue in this feedstock, as this issue has a dedicated purpose.)
